Feature/omri

OmriA opened the pull request.

OmriA committed:
Implemented TODOs in the code

Implement fields_info fucntion in arp class
Implement deserialize function in layer class

OmriA committed:
Router's basic ARP functionality

Added ipaddress field
Added support of squared brackets to search layers
Implemented RouterLeg class
Implemented IPLayer class
Implemented router parser for ethernet layer and arp layer
Implemented router response to arp requests

OmriA committed:
Router answers to ping sent to him

Implemented IcmpLayer class in l3
Added checksum calculation to IpLayer
Added responding to icmp requests

OmriA committed:
Bug fixes

Changed checking if variable has an instance to explicit check.
Now the router does not change the length field (not supposed to).
Added RawLayer class which contains all the data that we didn't parse.

OmriA committed:
Forwarding packets between networks

Basic implemenation of packet forwarding between networks

nivbromberg reviewed (changes requested):
Hi,
Please make this PR into `feature/skeleton`, as this is where you branched from.
This makes us PR only your changes and diff 😃

nivbromberg commented on src/pascy/l2.py:
```diff
@@ -12,15 +13,26 @@ class ArpLayer(Layer):
+                MacAddress("sender_hardware_addr"),
+                IPAddress("sender_protocol_addr"),
+                MacAddress("target_hardware_addr"),
+                IPAddress("target_protocol_addr")]
```
It is recommended to add a comma (`,`) at the end of any list item, including the last one.
This way, when someone will add support to a new protocol, the git blame won't be on him for adding the comma, but only for the relevant line

nivbromberg commented on src/pascy/l2.py:
```diff
@@ -12,15 +13,26 @@ class ArpLayer(Layer):
+
     SUB_LAYERS = [
         [ArpLayer, "ether_type", 0x806],
+        [IpLayer, "ether_type", 0x800]
```
As said before, adding a comma is recommended, as you may well see in this case 🙃

nivbromberg commented on src/pascy/l3.py:
```diff
@@ -0,0 +1,64 @@
+        return [UnsignedByte("type"),
+                UnsignedByte("code"),
+                UnsignedShort("checksum"),
+                ByteString("rest", 4)]
```
What is rest?
Did you mean payload?

nivbromberg commented on src/pascy/l3.py:
```diff
@@ -0,0 +1,64 @@
+                IPAddress("src"),
+                IPAddress("dst")]
+
+    def calc_checksum(self):
```
If this is similar (or even identical) to ICMP checksum, why not make both use the same function?
Also, consider `lru_cache`ing this function, as identical buffers will result in identical checksum

OmriA replied:
I am not sure if `lru_cache` is a smart idea here.
I combined the functions

nivbromberg replied:
If you encounter the same buffer, the checksum will be identical.
Consider using that fact into optimizing your code's performance

OmriA replied:
In second thought you are right.
I will change it.

I forgot the checksum is calculating only the header fields without the data itself

nivbromberg commented on src/pascy/layer.py:
```diff
@@ -106,14 +114,15 @@ def deserialize(self, buffer: bytes):
         """
-        # TODO: Implement this :)
-        pass
+        for _, field in self.fields.items():
```
Very nice and clean function!

nivbromberg commented on src/pascy/raw_layer.py:
```diff
@@ -0,0 +1,10 @@
+from pascy.fields import ByteString
+
+class RawLayer(Layer):
+
```
Why this extra line?

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+
+ROUTER_LEGS = []
+
+ARP_TABLE = {'1.1.1.2':'02:42:01:01:01:02',
```
What about MAC addresses of the router itself?

OmriA replied:
This is the leg's MAC

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+ARP_TABLE = {'1.1.1.2':'02:42:01:01:01:02',
+             '2.2.2.2':'02:42:02:02:02:02'}
+
+BROADCAST = 'FF:FF:FF:FF:FF:FF'
```
Isn't that also defined in `l2.py`? 🙃

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+
+BROADCAST = 'FF:FF:FF:FF:FF:FF'
+
+def get_packet(leg : RouterLeg) -> pascy.Layer:
```
The convention is not to put space **before** the colon, like this:
```python
def get_packet(leg: RouterLeg) -> pascy.Layer:
```

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+def get_packet(leg : RouterLeg) -> pascy.Layer:
+    # Recieve packet
+    raw_socket = leg.raw_socket
+    pkt_raw = raw_socket.recvfrom(65536)[0]
```
What is `65536`?

OmriA replied:
Buffer I can read

nivbromberg replied:
`65536` is a number, not a buffer...
Please explain yourself more clearly.

Also, make this a constant 🙃 
It's nice to have a file with all constant (usually named `const.py`), as this makes it easier to modify all constants throughout the project.
This is especially useful if you use them in many places.

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+def get_packet(leg : RouterLeg) -> pascy.Layer:
+    # Recieve packet
+    raw_socket = leg.raw_socket
+    pkt_raw = raw_socket.recvfrom(65536)[0]
```
Why `recvfrom`?
You take `[0]` anyway...

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+    dst = pascy.fields.MacAddress.mac2str(eth.dst)
+    # Checks if the packet is for me
+    if dst != leg.mac and dst != BROADCAST:
+        return None
```
What is the difference between `return` and `return None`?
Why do you explicitly return `None`?

OmriA replied:
Conventions.
If the function not supposed to return anything I will use return to stop it.
Here the function supposed to return something

nivbromberg replied:
But why not simply `return`?

OmriA replied:
Because I expect the function to return something (while calling it)

nivbromberg replied:
What does `return` without specifying anything do?

OmriA replied:
Returns None

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+
+    # Parse ip layer
+    elif eth.ether_type == pascy.EthernetLayer.IPV4_ETHER_TYPE:
+        ip = pascy.l3.IpLayer()
```
Wouldn't it make more sense to split parsing different layers to functions?
Something like:
```python
if eth.ether_type == pascy.EthernetLayer.ARP_ETHER_TYPE:
    parse_arp(eth, pkt_raw)

elif eth.ether_type == pascy.EthernetLayer.IPV4_ETHER_TYPE:
    parse_ip(eth, pkt_raw)
```
This way, `parse_ip` could also handle ICMP, making the code more modular and readable

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+
+        eth.connect_layer(ip)
+
+    if len(pkt_raw) > 0:
```
What does this do?
Document and explain

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+    return eth
+
+def response_arp(pkt, leg : RouterLeg):
+    resp_pkt = pascy.l2.EthernetLayer() / pascy.l2.ArpLayer() / pascy.RawLayer()
```
Why do you add raw?

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+    resp_pkt["ARP"].target_hardware_addr = pkt["Ethernet"].src
+    resp_pkt["ARP"].target_protocol_addr = pkt["ARP"].sender_protocol_addr
+
+    leg.raw_socket.sendall(resp_pkt.build())
```
Why `sendall`?

OmriA replied:
Because sendall throws an exception if something failed and responsible to send the whole pkt (if the packet is too big to be sent in one send call)

nivbromberg replied:
`send` doesn't throw such exceptions?

OmriA replied:
https://stackoverflow.com/questions/34252273/what-is-the-difference-between-socket-send-and-socket-sendall

nivbromberg replied:
Well this thread explains it differently than you did...
It says that `send` may send part of the buffer, while `sendall` throws an exception if it failed to send all buffer.

OmriA replied:
I said " and responsible to send the whole pkt (if the packet is too big to be sent in one send call)" :)

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+    dst_ip = socket.inet_ntoa(pkt["IP"].dst)
+    pkt["Ethernet"].src = leg.mac
+    pkt["Ethernet"].dst = ARP_TABLE[dst_ip]
+    pkt["IP"].ttl -= 1
```
_Nice_

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+        if leg.raw_socket == s:
+            return leg
+
+    return None
```
Why do you explicitly do this?
What happens if you don't state this? 🙃

OmriA replied:
What do you mean?

nivbromberg replied:
What happens if you don't write `return None`?
What does the function do?

OmriA replied:
It will act the same but I prefer doing it explicitly to be more readable

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+@lru_cache()
+def get_leg_by_ip(ip) -> RouterLeg:
+    for leg in ROUTER_LEGS:
+        if leg.subnet in ip:
```
Shouldn't that be the opposite?
IP address is inside a subnet 🤔

OmriA replied:
Yep.
But I used the subnet as part of the ip
e.g IP: 26.32.1.5. Subnet: 26.32.1

nivbromberg replied:
Oh you treated these as strings 👎 
This works well when the subnet mask is 24.
But what if it's not?
Also, it's bad practice to do this anyway.
Use `ipaddress` 🤗

OmriA replied:
Yeah, I know...
Tried to finish it as fast as possible and I would take care about it after the basic is working

nivbromberg replied:
Ok, fair enough.
Change that if you'll have the time 🤗

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+    return None
+
+def main():
+    ROUTER_LEGS.append(RouterLeg('net1', '1.1.1', '02:42:05:66:DC:60', '1.1.1.1'))
```
What is `1.1.1`?
Did you mean `1.1.1.0/24`?
You should explicitly state the subnet mask.
Many routing options differ in subnets

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,159 @@
+                forward_packet(pkt, leg)
+
+        for s in exceptional:
+            legs_sockets.remove(s)
```
Interesting.
Maybe log as well?

nivbromberg reviewed (changes requested):
Overall really good code!
One could see that you have really learned about networks and routing.
Your parsing library is very nice and really acts like the real `scapy`! 🥳 

However,
Please document and add explanations on complex/non-trivial logic.
Consider splitting some logic into functions

OmriA committed:
Esthetic commit

Added comma in the last element in every list
Splitted the parse of every layer to seperate functions
Combined l3 checksum calculation to one function
Added const.py file for constants in the code
Added log print when socket failed

nivbromberg commented on src/pascy/l3.py:
```diff
@@ -0,0 +1,60 @@
+from functools import lru_cache
+
+@lru_cache()
+def calc_checksum(bytes_buffer: bytes):
```
_Nice_

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,171 @@
+
+    dst = pascy.fields.MacAddress.mac2str(eth.dst)
+    # Checks if the packet is for me
+    if dst != leg.mac and dst != pascy.l2.MAC_BROADCAST:
```
Consider
```python
if dst not in [leg.mac, pascy.l2.MAC_BROADCAST]:
```
But whatever is more comfortable for you

nivbromberg commented on src/router.py:
```diff
@@ -0,0 +1,171 @@
+        for s in exceptional:
+            leg = get_leg_by_socket(s)
+            legs_sockets.remove(s)
+            print("Critical: lost connection with leg {}".format(leg.interface))
```
Very _nice_

nivbromberg reviewed (approved):
Very nice and clean code!
Good work!!

Please see comments and try to learn from them to the future 😄
